plugins/bufferlines/barbar.nix

```diff
@@ -69,249 +24,567 @@ let
+    ]
+    [
+      "hide"
+      "visible"
+    ]
+    "highlightAlternate"
+    "highlightInactiveFileIcons"
+    "highlightVisible"
+    [
+      "icons"
+      "bufferIndex"
+    ]
+    [
+      "icons"
+      "bufferNumber"
+    ]
+    [
+      "icons"
+      "button"
+    ]
+    [
+      "icons"
+      "filetype"
+      "customColors"
+    ]
+    [
+      "icons"
+      "separator"
+      "left"
+    ]
+    [
+      "icons"
+      "separator"
+      "right"
+    ]
+    "insertAtStart"
+    "insertAtEnd"
+    "maximumPadding"
+    "minimumPadding"
+    "maximumLength"
+    "semanticLetters"
+    "letters"
+    "sidebarFiletypes"
+    "noNameTitle"
+    "tabpages"
+  ];
+  imports =
+    let
+      basePluginPath = [
+        "plugins"
+        "barbar"
+      ];
+      settingsPath = basePluginPath ++ [ "settings" ];
+    in
+    [
+      (mkRemovedOptionModule (
+        basePluginPath
+        ++ [
+          "keymaps"
+          "silent"
+        ]
+      ) "Keymaps will be silent anyways. This option has always been useless.")
+      (mkRenamedOptionModule (basePluginPath ++ [ "excludeFileTypes" ]) (
+        settingsPath ++ [ "exclude_ft" ]
+      ))
+      (mkRenamedOptionModule (basePluginPath ++ [ "excludeFileNames" ]) (
+        settingsPath ++ [ "exclude_name" ]
+      ))
+      (mkRemovedOptionModule (
+        basePluginPath
+        ++ [
+          "icons"
+          "diagnostics"
+        ]
+      ) "Use `settings.icons.diagnostics` instead, but pay attention as the keys have changed.")
+      (mkRenamedOptionModule
+        (
+          basePluginPath
+          ++ [
+            "icons"
+            "filetype"
+            "enable"
+          ]
+        )
+        (
+          settingsPath
+          ++ [
+            "icons"
+            "filetype"
+            "enabled"
+          ]
+        )
+      )
+    ]
+    ++ (map
+      (
+        name:
+        mkRemovedOptionModule (
+          basePluginPath
+          ++ [
+            "icons"
+            name
+          ]
+        ) "Use `settings.icons.${name}` instead, but you should now use the real `snake_case` key names."
+      )
+      [
+        "alternate"
+        "current"
+        "inactive"
+        "modified"
+        "pinned"
+        "visible"
+      ]
+    );
+
+  extraOptions = {
+    keymaps = mapAttrs (
+      optionName: funcName:
+      helpers.mkNullOrOption (helpers.keymaps.mkMapOptionSubmodule {
+        defaults = {
+          mode = "n";
+          action = "<Cmd>Buffer${funcName}<CR>";
+        };
+        hasKey = true;
+        hasAction = true;
+      }) "Keymap for function Buffer${funcName}"
+    ) keymapsActions;
+  };
 
-    package = helpers.mkPluginPackageOption "barbar" pkgs.vimPlugins.barbar-nvim;
+  extraConfig = cfg: {
+    keymaps = filter (keymap: keymap != null) (
+      # TODO: switch to `attrValues cfg.keymaps` when removing the deprecation warnings above:
+      attrValues (filterAttrs (n: v: n != "silent") cfg.keymaps)
```

Comment:
I don't think the `filterAttrs` is needed, my understanding is that `mkRemovedOptionModule` uses an _assertion_, so `keymaps.silent` should never be defined.

Reply:
This is what I needed to do to solve https://github.com/nix-community/nixvim/pull/1608#issuecomment-2143498632

Reply:
Strange, but ok. I guess attempting to read the value here causes it to be evaluated? I figured those assertions would only affect _defining_ the option, not _reading_ it :/

Reply:
But the option does exist. It's hidden but it's there. So when you loop over `cfg.keymaps` it's there with its default.
So it is as it was used. Well it **is** used.

Reply:
I was under the impression the module system made a distinction between an option existing and a configuration being _defined_. 

I assumed the assertion was just for _defining_ the option, but I guess it's also for _reading_ its value. Makes sense, considering you'd want to throw if someone tries to read a removed option.

Just a shame it also affects reading the (entire) parent attrset, but I can't see how the module system would avoid that.